server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java

```diff
@@ -0,0 +1,484 @@
+        final ActivePeerFinder activePeerFinder = this.peerFinder.get();
+        assert activePeerFinder.running;
+        return activePeerFinder;
+    }
+
+    PeersResponse handlePeersRequest(PeersRequest peersRequest) {
+        synchronized (mutex) {
+            assert peersRequest.getSourceNode().equals(getLocalNode()) == false;
+            if (isActive()) {
+                final ActivePeerFinder activePeerFinder = getActivePeerFinder();
+                activePeerFinder.startProbe(peersRequest.getSourceNode().getAddress());
+                peersRequest.getKnownPeers().stream().map(DiscoveryNode::getAddress).forEach(activePeerFinder::startProbe);
+                return new PeersResponse(Optional.empty(), activePeerFinder.getKnownPeers(), currentTerm);
+            }
+        }
+
+        return onPeersRequestWhenInactive(peersRequest.getSourceNode());
+    }
+
+    public boolean isActive() {
+        synchronized (mutex) {
+            if (peerFinder.isPresent()) {
+                assert peerFinder.get().running;
+                return true;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    public void setCurrentTerm(long currentTerm) {
+        // Volatile write, no synchronisation required
+        this.currentTerm = currentTerm;
+    }
+
+    private DiscoveryNode getLocalNode() {
+        // No synchronisation required
+        final DiscoveryNode localNode = transportService.getLocalNode();
+        assert localNode != null;
+        return localNode;
+    }
+
+    @Override
+    protected void doStart() {
+        // No synchronisation required
+        executorService.set(executorServiceFactory.get());
+    }
+
+    @Override
+    protected void doStop() {
+        // No synchronisation required
+        ThreadPool.terminate(executorService.get(), 10, TimeUnit.SECONDS);
+    }
+
+    @Override
+    protected void doClose() {
+    }
+
+    /**
+     * Called on receipt of a PeersResponse from a node that believes it's an active leader, which this node should therefore try and join.
+     */
+    protected abstract void onMasterFoundByProbe(DiscoveryNode masterNode, long term);
+
+    /**
+     * Called on receipt of a PeersRequest when there is no ActivePeerFinder, which mostly means that this node thinks there's an
+     * active leader. However, there's no synchronisation around this call so it's possible that it's called if there's no active
+     * leader, but we have not yet been activated. If so, this should respond indicating that there's no active leader, and no
+     * known peers.
+     *
+     * @param sourceNode The sender of the PeersRequest
+     */
+    protected abstract PeersResponse onPeersRequestWhenInactive(DiscoveryNode sourceNode);
+
+    public interface TransportAddressConnector {
+        /**
+         * Identify the node at the given address and, if it is a master node and not the local node then establish a full connection to it.
```

Comment:
I think this is benign, but there are things we could do in the `TransportAddressConnector` to avoid this. For instance, if we are already connected to a non-master node then there's no need to identify it via a lightweight connection; perhaps we could cache responses too.

This PR is large enough without another implementation and test suite - the real implementation can come in a followup.